File: kernel/src/filesystem/ustar.cpp
/*
 * Copyright (c) 2026-present, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

#include "filesystem/ustar.hpp"

#include <ygg/string.hpp>

namespace ustar {

usize octal_to_binary(const u8 *array, usize size)
{
    usize count = 0;

    const u8 *character = array;
    while (size-- > 0) {
        count *= 8;
        count += *character - '0';
        ++character;
    }

    return count;
}

File lookup(const u8 *archive, const char *file_name)
{
    const u8 *ptr = archive;
    while (!memcmp(ptr + 257, "ustar", 5)) {
        const usize file_size = octal_to_binary(ptr + 0x7c, 11);
        if (!memcmp(ptr, file_name, strlen(file_name) + 1)) {
            return {
                .ptr = ptr + 512,
                .size = file_size,
            };
        }
        ptr += (((file_size + 511) / 512) + 1) * 512;
    }

    return {
        .ptr = nullptr,
        .size = 0,
    };
}

} // namespace ustar

File: kernel/src/filesystem/ustar.rs
//
// Copyright (c) 2026-present, SkillerRaptor
//
// SPDX-License-Identifier: MIT
//

fn octal_to_binary(data: &[u8]) -> u64 {
    let mut result = 0;

    for byte in data {
        result *= 8;
        result += (byte - b'0') as u64;
    }

    result
}

pub fn lookup<'a>(archive: &'a [u8], file_name: &str) -> Option<&'a [u8]> {
    let mut offset = 0;

    while archive.get(offset + 257..offset + 262) == Some(b"ustar") {
        let file_size = octal_to_binary(&archive[offset + 0x7c..offset + 0x7c + 11]);

        if archive.get(offset..offset + file_name.len()) == Some(file_name.as_bytes()) {
            return Some(&archive[offset + 512..offset + 512 + file_size as usize]);
        }

        offset += (((file_size as usize + 511) / 512) + 1) * 512;
    }

    None
}
